vespalib/src/vespa/vespalib/util/brain_float16.h

```diff
@@ -0,0 +1,104 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#pragma once
+
+#include <bit>
+#include <cstdint>
+#include <cstring>
+
+namespace vespalib {
+
+class BrainFloat16 {
+private:
+    uint16_t _bits;
+    struct TwoU16 {
+        uint16_t u1;
+        uint16_t u2;
+    };
+public:
+    constexpr BrainFloat16(float value) noexcept : _bits(float_to_bits(value)) {}
+    BrainFloat16() noexcept = default;
+    ~BrainFloat16() noexcept = default;
+    constexpr BrainFloat16(const BrainFloat16 &other) noexcept = default;
+    constexpr BrainFloat16(BrainFloat16 &&other) noexcept = default;
+    constexpr BrainFloat16& operator=(const BrainFloat16 &other) noexcept = default;
+    constexpr BrainFloat16& operator=(BrainFloat16 &&other) noexcept = default;
+
+    constexpr operator float () const noexcept { return bits_to_float(_bits); }
+
+    constexpr float to_float() const noexcept { return bits_to_float(_bits); }
+
+    constexpr void assign(float value) noexcept { _bits = float_to_bits(value); }
+
+    static constexpr uint16_t float_to_bits(float value) noexcept {
+        TwoU16 both{0,0};
+        static_assert(sizeof(TwoU16) == sizeof(float));
+        memcpy(&both, &value, sizeof(float));
+        if constexpr (std::endian::native == std::endian::big) {
+            return both.u1;
+        } else if constexpr (std::endian::native == std::endian::little) {
+            return both.u2;
+        } else {
+            return 0;
```

Comment:
consider static assert with error message here

Reply:
rewritten to allow static assert here and below.